spec/lib/random_example_spec.rb

```diff
@@ -25,6 +26,18 @@
       end
     end
 
+    it "returns the same output if a seed is detected" do
+      schema = GovukSchemas::Schema.random_schema(schema_type: "frontend")
+      # freeze time to avoid inconsistent `public_updated_at` values between runs
+      Timecop.freeze(Date.today) do
+        srand(777) # these srand calls would be in the upstream application
```

Comment:
These could impact other aspects of the test suite in their global form.

The approach you can probably take to avoid this is:

```
        previous_seed = srand(777)
        first_payload = GovukSchemas::RandomExample.new(schema: schema).payload
        second_payload = GovukSchemas::RandomExample.new(schema: schema).payload
        expect(first_payload).to eql(second_payload)
        srand(previous_seed)
```

Though this would be nicer in an around block if possible.

Reply:
Hmmm, I think this might give a false sense of security.

I see in your suggested code you've removed the `srand(777)` call between the `first_payload` and `second_payload` assignments. This call is necessary, as demonstrated below:

```ruby
srand(777)
rand(0..100) # 42
rand(0..100) # 91
# ... every `rand` call will produce a random value. So far, `srand` doesn't seem to have done anything useful, right?

# Let's call it a second time
srand(777)

# the rand calls return the same 'random' values in the same order as before.
rand(0..100) # 42
rand(0..100) # 91
```

Hopefully this demonstrates why the two `srand(777)` calls are necessary. It should also explain why calling `srand(previous_seed)` won't necessarily 'reset' the global state as we might expect.

Let's say our tests have randomly been given the seed 245178:

```ruby
srand(245178)

# ...and we make a few `rand()` calls...
rand(0..100) # 37
rand(0..100) # 88
```

Now let's say that between the two `rand` calls above, we now have the random_example_spec.rb test, which sets the `srand`:

```ruby
srand(777)
# ...tests etc
# At the end, we reset back to the original `srand`:
srand(245178)
```

Unless `srand` 'picks up where it left off' with the old 245178 value, that second `rand()` call is going to return 37, not 88 as expected. I don't know if that's better than not attempting to reset the `srand` at all 🤔 but can't think of a different way of testing it to what I currently have.

Reply:
Oh yes shoot I see what you mean, d'oh. Good catch, I didn't consider the reset aspect at all.

Hmm that is a bit tricky then to avoid other worldly side-effects. There is the approach @thomasleese suggested still which seems to offer an isolated approach, is that rather grim to implement?

Reply:
It should be possible, but means either:

1. An API change whereby we pass a `seed` parameter from upstream.
2. We retrieve the current `srand` value and use that to seed as `Random.new(current_srand_val)`.

2 would be preferred but I can't find a way of retrieving the `srand` value without calling it (and thereby changing it) - Schrödinger's srand 😅

Reply:
If not passed in explicitly, we could generate a seed by calling rand() rather than trying to find the global seed.

If we match the global seed, the class ends up being deterministic by default as the global seed is unlikely to change during the execution of an application, so each time the class is called, it would return the same result. I suspect that might be surprising behaviour when the seed is not passed in explicitly.

Reply:
Sounds like you're in favour of having a `seed` parameter option in the API, @thomasleese .

Shall I go that route, @kevindew ?

Reply:
It sounds good. Just to clarify as I'm not sure I've followed all the suggestion discussion. Is this an api change to https://github.com/alphagov/govuk_schemas/blob/764b2ac6be516a43c5a32bf6bb19ef57af6ba899/lib/govuk_schemas/random_example.rb#L34 which adds an additional seed argument?

And then change GovukSchemas::RandomExample to accept an instance of `Random` which is set with the second argument? I guess Random would need to stop acting as a collection of static functions and act as a class?

Reply:
OK, it was quite fiddly. Have now pushed up a few more commits.